### PR TITLE
- new approval type for address book management

### PR DESCRIPTION
**Pull request by bflood-strike**
Adds a new approval type for address book management

## Description

Adding/Removing entries to/from the address book and balance account whitelist updates can be performed in a single approval

## Motivation and Context

new feature

## How Has This Been Tested?

added new unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Bug fix breaking (fix that would cause existing functionality to change)
- [ X] New feature (non-breaking change which adds functionality)
- [ ] New feature breaking (change which adds functionality and causes existing functionality to change)

## Checklist:

- [X ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [X ] I have added tests to cover my changes.
- [X ] All new and existing tests passed.

### CODE REVIEW

**Comment by dgabriele** [comment]
👍🏼 looks good. all tests 🟢

**Review by dgabriele — APPROVED** [review]

File: src/instruction.rs [hunk 1]
```diff
@@ -823,6 +865,65 @@ impl WalletUpdate {
...
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct AddressBookUpdate {
+    pub add_address_book_entries: Vec<(SlotId<AddressBookEntry>, AddressBookEntry)>,
+    pub remove_address_book_entries: Vec<(SlotId<AddressBookEntry>, AddressBookEntry)>,
+    pub balance_account_whitelist_updates: Vec<BalanceAccountWhitelistUpdate>,
+}
+
+impl AddressBookUpdate {
+    fn unpack(bytes: &[u8]) -> Result<AddressBookUpdate, ProgramError> {
+        if bytes.len() < 7 {
```
Comment (bholzman) [inline src/instruction.rs:904]: we don't need this, do we?

File: src/model/multisig_op.rs [hunk 2]
```diff
@@ -679,6 +683,21 @@ impl MultisigOpParams {
...
+            } => {
+                let mut update_bytes: Vec<u8> = Vec::new();
+                update.pack(&mut update_bytes);
+
+                let mut bytes: Vec<u8> = Vec::new();
+                bytes.resize(1 + PUBKEY_BYTES + update_bytes.len(), 0);
+                bytes[0] = 10; // type code
+                bytes[1..1 + PUBKEY_BYTES].copy_from_slice(&wallet_address.to_bytes());
+                bytes[1 + PUBKEY_BYTES..1 + PUBKEY_BYTES + update_bytes.len()]
+                    .copy_from_slice(&update_bytes);
```
Comment (bholzman) [inline src/model/multisig_op.rs:698]: not specific to this PR, but there are a bunch of ops that have this "type code", "wallet address", "some update in bytes" structure, might be worth DRYing it a bit?

File: src/error.rs [hunk 3]
```diff
@@ -89,6 +89,11 @@ pub enum WalletError {
     UnknownSigner,
     #[error("DApp Not Allowed")]
     DAppNotAllowed,
+
+    // 25
+    /// Cannot remove an address from address book since it is whitelisted on a balance account.
```
Comment (bholzman) [inline src/error.rs:94]: I'm about to create a PR that does a bit more reworking of these error codes, and actually renames `SlotAlreadyInUse` to `DestinationInUse`

File: src/model/wallet.rs [hunk 4]
```diff
@@ -586,6 +609,11 @@ impl Wallet {
             msg!("Failed to enable transfer destinations: address book does not contain one of the given destinations");
             return Err(WalletError::InvalidSlot.into());
         }
+        let balance_account = &mut self.balance_accounts[balance_account_index].borrow_mut();
+        if !destinations.is_empty() && balance_account.is_whitelist_disabled() {
```
Comment (bholzman) [inline src/model/wallet.rs:613]: nice catch

**Review by bholzman — COMMENTED** [review]
Looks great!

File: src/model/multisig_op.rs (hunk 2 above)
Comment (bflood-strike) [reply src/model/multisig_op.rs]: yep, updated to have these updates calls a common function

File: src/instruction.rs (hunk 1 above)
Comment (bflood-strike) [reply src/instruction.rs]: yep removed

File: src/error.rs (hunk 3 above)
Comment (bflood-strike) [reply src/error.rs:94]: ok, I'll hold off til that is merged and then use the DestinationInUse as oppose to the new AddressInUse I added.

Comment (bholzman) [reply src/error.rs:94]: my PR is merged fyi

Comment (bflood-strike) [reply src/error.rs]: ok rebased and adjusted to use new error codes

**Review by bholzman — APPROVED** [review]

**Pull request merged**